### PR TITLE
raft: add progress CatchUp

### PR DESCRIPTION
**Pull request by yichengq**
CatchUp marks whether the remote node catches up with the leader. If it
does, leader can send MsgApp optimistically to the follower.

It comes into CatchUp status when leader receives a valid MsgAppResp
from the follower.

When it is bootstrapped or rejected by the remote, CatchUp is set to
false, and it falls back to slow replication method.

/cc @bdarnell

### CODE REVIEW

**Comment by yichengq** [comment]
We are doing integration test: kill one member for a long time and expect it will recover from incoming snapshot.

Based on current implementation, the member has been matched with the leader, and is in the optimistic update path always. It sends out hundreds of MsgApp optimistically in one RTT, which are rejected due to restart or message lost, then triggers several hundreds times of MsgApp starting from match+1 at the same time. It is horrible because it can eat up tons of memory.

CatchUp method makes it fall back to slow replication way when rejection, as what it does at bootstrap stage. It catches up the leader steadily, and recover to optimistic update when catched up again.

File: raft/raft.go [hunk 1]
```diff
@@ -52,6 +52,12 @@ func (st StateType) String() string {
 type Progress struct {
 	Match, Next uint64
 	Wait        int
+	// If the last MsgAppResp updates the Match, which indicates that the
+	// follower has caught up the progress, CatchUp will be set.
+	// If a valid MsgApp with rejection is received, CatchUp will be unset.
+	// When CatchUp is set, leader can send MsgApp optimistically to the
+	// follower.
+	CatchUp bool
```
Comment (xiang90) [inline raft/raft.go]: how about just Matched? The previous problem is that "Match!=0" is not a good indication of a Match.

**Comment by xiang90** [comment]
@yichengq This makes sense. But I would like to see a separate test just for the new field. And we probably should combine the `Unreachable` , `PendingSnapshot` , `Conflict` and `OK` (conflict, ok is the catchup in this commit) fields into one. The progress should only be one of the state.

File: raft/raft.go (hunk 1 above)
Comment (xiang90) [inline raft/raft.go]: or maybe use a negative one `Conflicting`

**Comment by xiang90** [comment]
@yichengq Basically, we should only enable optimistic sending when the last appendMessage response is an `accept`. Right?

**Comment by xiang90** [comment]
@bdarnell After an offline discussion with @yichengq , we realized that we should take a more carefully way to deal with this. We are going to treat the `Progress` as a state machine, and work out the full transition graph before writing any code around this. Will share with you the graph soon.

**Pull request closed without merging**